### PR TITLE
Stats: Add iteration and sink predicates for histograms

### PR DESCRIPTION
**Pull request by pradeepcrao**
Commit Message:
Additional Description:
This is a followup to #18805 that adds the ability to filter histograms to be flushed to sink. Note that this only affects which histograms are flushed and does not change which histograms are merged during every flush operation.

Risk Level: Low
Testing: Added Tests
Docs Changes:  N/A
Release Notes: included.
Platform Specific Features: N/A

### CODE REVIEW

File: test/common/stats/thread_local_store_test.cc
Comment (jmarantz) [inline test/common/stats/thread_local_store_test.cc:1617]: why not just EXPECT_EQ on this? We typically don't have this `(void)` cast in Envoy code for whatever reason.

File: source/server/server.cc
Comment (jmarantz) [inline source/server/server.cc:193]: can we verify what happens for histograms that are never sinked but where we call recordValue regularly?

Do they consume unbounded memory? Maybe make a unit test that runs an infinite loop with a bunch of recordValue() and see if it consumes increasing memory while you let it spin.

If so, then this might be a memory leak.

This case is also why I wanted to do histograms separate from the other stuff, as we'll need to figure out how to deal with this.

**Review by jmarantz — COMMENTED** [review]
looks great modulo some minor nits and one concern about histogram memory growth.

File: test/common/stats/thread_local_store_test.cc
Comment (jmarantz) [inline test/common/stats/thread_local_store_test.cc:1612]: this is not a good pattern -- I realize this pretty much has to work, but officially I'm pretty sure it's undefined behavior to save a pointer held in a unique_ptr that you then use as a the source of a move.

Instead, can you provide an accessor: Store::sinkPredicates() and reference them from the store, and change the other test-cases that use this pattern also?

File: test/common/stats/thread_local_store_test.cc
Comment (jmarantz) [inline test/common/stats/thread_local_store_test.cc:1928]: note that there is already source/common/memory/stats.h -- will that do what you want?

I'm skeptical this will work on all non-windows platforms. Apple? Mobile?

File: test/common/stats/thread_local_store_test.cc
Comment (jmarantz) [inline test/common/stats/thread_local_store_test.cc:1965]: can we make these numbers much much higher? With such low numbers we might not need to allocate any new pages.

BTW I didn't necessarily expect permanently save a memory test. What I was thinking is that you'd have one that ran for like 10 minutes adding tons of data points to histograms, and look at the process with 'top' while it was running to see if it's increasing in memory.

**Review by jmarantz — COMMENTED** [review]
this looks great! Just a few nits.

**Comment by jmarantz** [comment]
/wait

also need to fix format.

**Comment by jmarantz** [comment]
Please let me know when this is ready for another look 

/wait

File: test/common/stats/thread_local_store_test.cc
Comment (pradeepcrao) [reply test/common/stats/thread_local_store_test.cc]: Changed the test to make sure that merge is called for histograms that are not included by sink predicates instead (since that's what we should be testing anyways).

File: test/common/stats/thread_local_store_test.cc
Comment (pradeepcrao) [reply test/common/stats/thread_local_store_test.cc]: removed this.

File: test/common/stats/thread_local_store_test.cc
Comment (pradeepcrao) [reply test/common/stats/thread_local_store_test.cc]: Added sinkPredicates() to StoreRoot.

File: test/common/stats/thread_local_store_test.cc
Comment (pradeepcrao) [reply test/common/stats/thread_local_store_test.cc]: Done

File: test/common/stats/thread_local_store_test.cc
Comment (pradeepcrao) [reply test/common/stats/thread_local_store_test.cc]: Done

File: test/common/stats/allocator_impl_test.cc
Comment (pradeepcrao) [reply test/common/stats/allocator_impl_test.cc]: Done

File: source/common/stats/isolated_store_impl.h
Comment (pradeepcrao) [reply source/common/stats/isolated_store_impl.h]: Done.

File: source/server/server.cc
Comment (pradeepcrao) [reply source/server/server.cc]: Removed the keyword.

**Comment by pradeepcrao** [comment]
> Please let me know when this is ready for another look
> 
> /wait

Hi Josh, the PR is ready for another look.

File: test/common/stats/allocator_impl_test.cc
Comment (jmarantz) [inline test/common/stats/allocator_impl_test.cc:418]: s/Stats::// here -- we are already in `Stats`

**Review by jmarantz — COMMENTED** [review]
Looks good basically -- still need to look over tests.

Can you  put into the description what you did to determine we won't leak memory with a long-running test? I know you did that earlier and ran into an issue with leaks due to mocks, but I'm not sure if that got resolved or how long a test you ran, etc.

File: test/common/stats/allocator_impl_test.cc
Comment (pradeepcrao) [reply test/common/stats/allocator_impl_test.cc:418]: done

**Comment by pradeepcrao** [comment]
> Looks good basically -- still need to look over tests.
> 
> Can you put into the description what you did to determine we won't leak memory with a long-running test? I know you did that earlier and ran into an issue with leaks due to mocks, but I'm not sure if that got resolved or how long a test you ran, etc.

Here are the results (First column is iteration number, and second column is increase in memory)
(The output is from here: https://github.com/envoyproxy/envoy/pull/19640/commits/cf185d5603d3b599a02992a6c7b065abdd4028c8#diff-3d1c3253b98e87621c3ecd349e647a1f1fe1c367e2d21ac9c6e6bb4042685632R2032)
1  -2912
10  -512
100  7168
1000  15360
10000  22016
100000  32256
1000000  36352

**Comment by jmarantz** [comment]
So there's some kind of leak?

I was expecting we could loop forever and the memory growth would be zero. Can we find out what we are still leaking?

**Comment by pradeepcrao** [comment]
> So there's some kind of leak?
> 
> I was expecting we could loop forever and the memory growth would be zero. Can we find out what we are still leaking?

I will investigate further, but do you feel that investigation is a blocker this PR?

**Comment by jmarantz** [comment]
Yeah we need to not leak memory.

I thought maybe (per my original suggestion) you could unblock most of it by splitting off the histogram-predicates into a smaller, high-risk PR, and leave the forEachHistogram API in a larger, low-risk PR.

**Comment by pradeepcrao** [comment]
> Yeah we need to not leak memory.
> 
> I thought maybe (per my original suggestion) you could unblock most of it by splitting off the histogram-predicates into a smaller, high-risk PR, and leave the forEachHistogram API in a larger, low-risk PR.

I was implying that the memory leak was unrelated to my PR (sorry I was not clear).

Indeed, I get the same behaviour with the test in a workspace with no other changes. (see: https://github.com/envoyproxy/envoy/pull/19654/files#diff-3d1c3253b98e87621c3ecd349e647a1f1fe1c367e2d21ac9c6e6bb4042685632R1918)

1  -2912
10  -512
100  7168
1000  15616
10000  22272
100000  30976
1000000  37888

Note that memory usage stops increasing if I reduce the range of distribution of random numbers to choose from for histogram values here https://github.com/envoyproxy/envoy/pull/19654/files#diff-3d1c3253b98e87621c3ecd349e647a1f1fe1c367e2d21ac9c6e6bb4042685632R1899

For eg, if I reduce the range to (0,2000), memory usage stays constant after 10000 merges:
1  -2912
10  -512
100  7168
1000  9344
10000  12800
100000  12800
1000000  12800

**Comment by jmarantz** [comment]
The test you have is good but we should find a way to run it without a leak (e.g. substitute real objects for mocks). I can help with that if needed.

Moreover, per IM, I think you can try to make a version with envoy-static that adds some histogram predicates, and run load over the coarse of minutes, looking at memory usage in 'top', and verifying that histograms are actually working despite not being sinked, by checking admin `/stats` manually. I think this shouldn't be *that* hard to do.

**Comment by pradeepcrao** [comment]
Hi @jmarantz , could you please take a second look at this PR? I modified a test to make sure we always merge all histograms even when we have SinkPredicates.

Once this is in, I'll investigate only merging histograms that are in the SinkPredicates.

**Review by jmarantz — COMMENTED** [review]
I think when this goes in, any application that has set up a predicate, but has not started calling the previously-non-existing 'includeHistogram' will see all their histogram reports dropped :)

We need to phase this in I think. I suspect the only user of sink predicates is us but we should verify, and for our own benefit I think you should add the includeHistogram in one PR, then we can start calling it in our applications, and *then* you can make it actually functional.

WDYT?

**Comment by pradeepcrao** [comment]
> I think when this goes in, any application that has set up a predicate, but has not started calling the previously-non-existing 'includeHistogram' will see all their histogram reports dropped :)
> 
> We need to phase this in I think. I suspect the only user of sink predicates is us but we should verify, and for our own benefit I think you should add the includeHistogram in one PR, then we can start calling it in our applications, and _then_ you can make it actually functional.
> 
> WDYT?

Would adding a runtime flag (off by default) that gates filtering the histograms make the transition smoother?

**Comment by jmarantz** [comment]
Yes a runtime flag could solve this also. Either way you'd need two PRs. Either:
 * one PR to add the new API, a second ot make the API work (done)
 * this PR with runtime augmentation (& doc), and then another one to remove the runtime flag after 6 months

I don't have a strong preference.

**Comment by jmarantz** [comment]
CI failure looks real:
```
[ RUN      ] HistogramParameterisedTestGroup/HistogramParameterisedTest.ForEachSinkedHistogram/EnableIncludeHistograms
/opt/llvm/bin/../include/c++/v1/__memory/unique_ptr.h:280:12: runtime error: reference binding to null pointer of type 'Envoy::Stats::ThreadLocalStoreImpl'
    #0 0x361d6db in std::__1::unique_ptr<Envoy::Stats::ThreadLocalStoreImpl, std::__1::default_delete<Envoy::Stats::ThreadLocalStoreImpl> >::operator*() const /opt/llvm/bin/../include/c++/v1/__memory/unique_ptr.h:280:5

```

File: test/common/stats/allocator_impl_test.cc
Comment (jmarantz) [reply test/common/stats/allocator_impl_test.cc:418]: Still doesn't look done. You need the TestUtil:: but I don't think you need the Stats::.

**Review by jmarantz — COMMENTED** [review]
/wait

thanks for doing this; getting close.

File: test/common/stats/stat_test_utility.h
Comment (jmarantz) [inline test/common/stats/stat_test_utility.h:225]: WDYT of exposing this as an add(StatName stat_name) and has(Statname stat_name) rather then exposing the rep by returning the member variable reference?

File: test/common/stats/thread_local_store_test.cc
Comment (jmarantz) [inline test/common/stats/thread_local_store_test.cc:1969]: any particular reason not to put this in the ctor? Adding code comments would be fine.

**Review by jmarantz — COMMENTED** [review]
lgtm with small nits. thank you!

It does need a release note I think.

File: test/common/stats/thread_local_store_test.cc
Comment (pradeepcrao) [reply test/common/stats/thread_local_store_test.cc:2036]: done.

File: test/common/stats/thread_local_store_test.cc
Comment (pradeepcrao) [reply test/common/stats/thread_local_store_test.cc:1969]: This needs store_ which is created in SetUp for the  base class. Added comments.

File: test/common/stats/stat_test_utility.h
Comment (jmarantz) [inline test/common/stats/stat_test_utility.h:228]: nit: state here that the StatName's backing store (e.g. StatNamePool) needs to live longer than the TestSinkPredicates object.

File: test/common/stats/thread_local_store_test.cc
Comment (jmarantz) [inline test/common/stats/thread_local_store_test.cc:2014]: I see what you are doing here and I think this is the right flow. But maybe call this test_sink_predicates? They are not moved yet :)

File: test/common/stats/thread_local_store_test.cc
Comment (jmarantz) [inline test/common/stats/thread_local_store_test.cc:217]: can you rename this testSinkPredicatesOrDie() or something?

File: test/common/stats/thread_local_store_test.cc
Comment (jmarantz) [inline test/common/stats/thread_local_store_test.cc:2091]: actually I feel like this pattern has UB in a sense, because the store_ which owns the predicates, will be destroyed after the pool, and contains a StatNameHashSet that depends on the pool.

I think this problem was pre-existing, and benign because flat_hash_map's destructor must not need to call hash or equals functions on the keys. But I think it'd be better to put the pool into the TestSinkPredicates. WDYT?

**Review by jmarantz — COMMENTED** [review]
lgtm; A few naming nits and a pedantic lifetime debate remain :)

File: test/common/stats/stat_test_utility.h
Comment (pradeepcrao) [reply test/common/stats/stat_test_utility.h:228]: Done.

File: test/common/stats/thread_local_store_test.cc
Comment (pradeepcrao) [reply test/common/stats/thread_local_store_test.cc:2014]: Done.

File: test/common/stats/thread_local_store_test.cc
Comment (pradeepcrao) [reply test/common/stats/thread_local_store_test.cc:217]: Done.

File: test/common/stats/thread_local_store_test.cc
Comment (pradeepcrao) [reply test/common/stats/thread_local_store_test.cc]: It was just easier to declare  pool_ in the test class before store_ than making the predicates own the pool and making that work with allocator_impl_test. Please let me know if that's acceptable.

**Review by jmarantz — DISMISSED** [review]
looks great, thanks!

/assign-from @envoyproxy/senior-maintainers

**Comment by jmarantz** [comment]
needs format update

**Comment by jmarantz** [comment]
needs non-google senior maintainer approval.

sorry Pradeep -- this does need a release note also.

**Review by jmarantz — DISMISSED** [review]

**Review by jmarantz — APPROVED** [review]

**Comment by jmarantz** [comment]
@ggreenway will you be able to look at this?

File: source/common/runtime/runtime_features.cc
Comment (ggreenway) [inline source/common/runtime/runtime_features.cc:96]: Why is a FALSE_RUNTIME_GUARD instead of a RUNTIME_GUARD?

**Comment by jmarantz** [comment]
@pradeepcrao were you going to pick this up again or should this remain closed?

**Comment by pradeepcrao** [comment]
I missed Greg's comment. Will address this now.

File: source/common/runtime/runtime_features.cc
Comment (pradeepcrao) [reply source/common/runtime/runtime_features.cc:96]: This disables filtering the histograms by default, so envoy users using sink predicates will not experience histograms silently not being flushed to sinks.

**Comment by jmarantz** [comment]
CI is failing. Also is the release prediction (1.27) still accurate?

/wait

**Comment by pradeepcrao** [comment]
> CI is failing. Also is the release prediction (1.27) still accurate?
> 
> /wait

Qualifying the fix for the CI failure.

Given that anyone who has implemented `SinkPredicates` will get a build failure that they need to react to (override `includeHistogram`), roughly 2 releases (1.25 and 1.26) should be sufficient time to enable the behaviour by default. What do you think?

**Review by jmarantz — APPROVED** [review]
@yurykats @stevenzzzz FYI

This still lgtm :)

**Pull request merged**
